cpp/src/tests/snmg_pagerank/snmg_pagerank_test.cu

```diff
@@ -0,0 +1,280 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "gtest/gtest.h"
+#include "high_res_clock.h"
+#include "cuda_profiler_api.h"
+#include <cugraph.h>
+#include <omp.h>
+#include "test_utils.h"
+#include "snmg_test_utils.h"
+#include "snmg/pagerank.cuh"
+
+//#define SNMG_VERBOSE
+
+
+typedef struct MGPagerank_Usecase_t {
+  std::string matrix_file;
+  std::string result_file;
+
+  MGPagerank_Usecase_t(const std::string& a, const std::string& b) {
+    // assume relative paths are relative to RAPIDS_DATASET_ROOT_DIR
+    // if RAPIDS_DATASET_ROOT_DIR not set, default to "/datasets"
+    const std::string& rapidsDatasetRootDir = get_rapids_dataset_root_dir();
+    if ((a != "") && (a[0] != '/')) {
+      matrix_file = rapidsDatasetRootDir + "/" + a;
+    } else {
+      matrix_file = a;
+    }
+    if ((b != "") && (b[0] != '/')) {
+      result_file = rapidsDatasetRootDir + "/" + b;
+    } else {
+      result_file = b;
+    }
+  }
+  MGPagerank_Usecase_t& operator=(const MGPagerank_Usecase_t& rhs) {
+    matrix_file = rhs.matrix_file;
+    result_file = rhs.result_file;
+    return *this;
+  }
+} MGPagerank_Usecase;
+
+class Tests_MGPagerank : public ::testing::TestWithParam<MGPagerank_Usecase> {
+  public:
+  Tests_MGPagerank() {  }
+  static void SetupTestCase() {  }
+  static void TearDownTestCase() { }
+  virtual void SetUp() {  }
+  virtual void TearDown() {  }
+
+  static std::vector<double> mgpr_time;   
+  template <typename val_t>
+  void verify_pr(gdf_column* col_pagerank, const MGPagerank_Usecase& param){
+    // Check vs golden data
+    if (param.result_file.length()>0)
+    {
+      int m = col_pagerank->size;
+      std::vector<val_t> calculated_res(m);
+      CUDA_RT_CALL(cudaMemcpy(&calculated_res[0],   col_pagerank->data, sizeof(val_t) * m, cudaMemcpyDeviceToHost));
+      std::sort(calculated_res.begin(), calculated_res.end());
+      FILE* fpin = fopen(param.result_file.c_str(),"rb");
+      ASSERT_TRUE(fpin != NULL) << " Cannot read file with reference data: " << param.result_file << std::endl;
+      std::vector<val_t> expected_res(m);
+      ASSERT_EQ(read_binary_vector(fpin, m, expected_res), 0);
+      fclose(fpin);
+      val_t err;
+      int n_err = 0;
+      for (int i = 0; i < m; i++) {
+          //if(i > (m-10))
+          //  std::cout << expected_res[i] << " " << calculated_res[i] <<std::endl;
+          err = fabs(expected_res[i] - calculated_res[i]);
+          if (err> 1e-5)
+          {
+              n_err++;
+          }
+      }
+      if (n_err) {
+          //EXPECT_NEAR(tot_err/n_err, cugraph_Const<T>::tol, cugraph_Const<T>::tol*9.99); // Network x used n*1e-10 for precision
+          EXPECT_LE(n_err, 0.001*m); // we tolerate 0.1% of values with a litte difference
+          //printf("number of incorrect entries: %d\n", n_err);
```

Comment:
Is this debugging code? You could use ifdefs here.

Reply:
Agreed but will be addressed in the next PR. This file is WIP and will be finalized along with the rest of the Pagernak solver. I'd rather do all the cleaning at once.